### PR TITLE
Add HTTP request-count regression tests for product upsert and status refresh

### PR DESCRIPTION
**Pull request by joemcgill**
Resolves [GOOWOO-1036](https://linear.app/a8c/issue/GOOWOO-1036/add-http-request-count-regression-tests-for-mapi-product-sync)

## Summary

Following up on the recent WCS/Merchant API traffic incident (per-product status GETs were ~87% of Merchant API calls, a 4.8x increase over the pre-migration baseline, fixed by #3779), this adds a fixed regression suite that asserts outbound HTTP *request counts* for routine sync tasks stay proportional to batch/page size rather than to catalog size.

- **`ProductUpsertRequestBudgetTest`** — mocks `MerchantApiClient` directly beneath `MapiProductInputsService` and asserts `insert_many()`/`delete_many()` issue `ceil(N / batch_size)` `batch_async()` calls for N = 1 through 2500, and never fall back to per-item `post`/`patch`/`get`/`delete`/`request(_async)` calls.
- **`ProductStatusRefreshRequestBudgetTest`** — drives `UpdateMerchantProductStatuses` through a full multi-page refresh (up to a simulated 10,000-product catalog) by having the mocked `schedule_immediate()` synchronously re-fire the job's action hook, the same way separate Action Scheduler runs carry the page token in production. Asserts `list_page()` is called exactly `ceil(N / page_size)` times and that `get()`/`get_many()` are never called — a direct regression test for the incident itself.
- **`DataSourceResolutionRequestBudgetTest`** — the two suites above stub `MapiDataSourcesService` out entirely, so they say nothing about the traffic `ensure_data_source_for()` generates on its own, even though it's called once per input inside both write loops. This suite wires the *real* `MapiDataSourcesService` (backed by a mocked `MerchantApiClient`) into a real `MapiProductInputsService` and asserts resolution requests (`dataSources.get`/`.list`/`.post`) scale with the number of distinct `(contentLanguage, feedLabel)` pairs touched, not with product count — and that both the persisted option cache and the in-memory per-run cache each independently keep that flat.

All three mock at the layer where the corresponding fan-out decision actually lives in code (the service for product writes, the job for status reads, the resolver for data-source lookups), reusing the existing mocking conventions already used in `MapiProductInputsServiceTest`, `UpdateMerchantProductStatusesTest`, and `MapiDataSourcesServiceTest`. A small shared `RequestBudgetTestCase` base class holds the fixtures common to more than one of these suites.

This covers three of five planned "request budget" regression suites. The remaining two (auth-failure short-circuit, coupon/shipping-region loops) are being added to this same PR rather than split into separate ones.

## Test plan

- [x] `php -l` on all new files
- [x] `./vendor/bin/phpcs --standard=tests/phpcs.xml.dist` clean on all new files
- [x] CI `PHP Unit Tests` workflow passes across the full PHP/WooCommerce version matrix

### CODE REVIEW

**Review by joemcgill — COMMENTED** [review]
## 🤖 AI Code Review

**Issue:** No linked tracker issue — PR description serves as acceptance criteria
**Branch:** `add/mapi-request-budget-tests` → `develop`

### Summary

The three new test files cover the stated objectives well: the `ceil(N / batch_size)` and `ceil(N / page_size)` math is correct, boundary cases are solid, and the `schedule_immediate` synchronous-replay trick is a faithful simulation of production Action Scheduler paging. The `MAX_RESCHEDULES` circuit breaker is a thoughtful fix for the hang risk the PR describes. However, the diff contains a third file (`DataSourceResolutionRequestBudgetTest.php`) that the PR title and description say is a planned follow-up, creating a meaningful scope mismatch, and two minor quality issues (dead constant, duplicate helper) are worth addressing.

### 🟡 Suggestions

- **Undescribed third file** (`tests/Unit/API/Google/Mapi/RequestBudget/DataSourceResolutionRequestBudgetTest.php`): The PR description explicitly says "follow-ups (data-source resolution volume, …) are planned as separate PRs," yet the diff adds 311 lines for exactly that. The file is well-written and logically fits the suite, but the PR description as written tells a reviewer this test isn't here. Either update the description to cover all three files (explaining the caching layers it exercises), or split the file to a follow-up PR as planned. This matters mainly for reviewability and changelog clarity, not correctness.

- **`respond_ok_to_every_sub_request` duplication** (`ProductUpsertRequestBudgetTest.php` line 643, `DataSourceResolutionRequestBudgetTest.php` line 250): The two methods are byte-for-byte identical. A thin shared base class (`RequestBudgetTestCase extends UnitTest`) in the `RequestBudget\` directory would eliminate the duplication and give a natural home for any additional shared fixtures as the suite grows.

### 🔵 Nitpicks

- (`ProductStatusRefreshRequestBudgetTest.php` line 355): `PROCESS_ITEM_HOOK` is defined but never referenced anywhere in this class — the hook value is correctly propagated through `$this->job->init()` and `schedule()` at runtime, so the constant is dead code. Remove it.

- (`ProductStatusRefreshRequestBudgetTest.php`): The `budget_sizes` data-provider name collides with the identically-named provider in `ProductUpsertRequestBudgetTest`. No runtime problem (different classes), but renaming to something like `page_configurations` would make `--filter` targeting less ambiguous.

- (`DataSourceResolutionRequestBudgetTest.php` lines 203–212, 222–241): Two tests set `get`/`post` expectations but never call `assert_client_never_calls_a_per_item_write_method()`, unlike the data-provider tests. Minor consistency gap.

---
*This review was generated by Claude. A human review is still required before merging.*

File: tests/Unit/API/Google/Mapi/RequestBudget/ProductUpsertRequestBudgetTest.php [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+		$this->client->expects( $this->never() )->method( 'request' );
+		$this->client->expects( $this->never() )->method( 'request_async' );
+	}
+
+	protected function service_batch_size(): int {
+		$method = new ReflectionMethod( MapiProductInputsService::class, 'get_batch_size' );
+		$method->setAccessible( true );
+
+		return $method->invoke( $this->service );
+	}
```
Comment (eason9487) [inline tests/Unit/API/Google/Mapi/RequestBudget/ProductUpsertRequestBudgetTest.php]: Since the expected batch count is computed from the service's current `get_batch_size()` via reflection, a regression of the default (for example, 100 down to 10) would recompute the expectation and still pass, doubling batch POSTs with nothing failing anywhere.

**Review by eason9487 — APPROVED** [review]
Thanks for the work!

This raises the odds of catching regressions on these code paths: the refresh suite guards the exact shape of the incident (a multi-page refresh that never issues per-product reads), the upsert suite pins the relationship between input count and batch count, and the resolver suite covers both cache layers separately.

One suggestion to consider alongside: removing `MapiProductsService::get()` and `get_many()` altogether. After #3779, nothing outside the connection-test debug page calls them, so they appear to be unused outside that page now.
